docs/_docs/31-fasterkv-revivification.md

```diff
@@ -0,0 +1,68 @@
+---
+title: "FasterKV Manual Locking"
+permalink: /docs/fasterkv-revivification/
+excerpt: "FasterKV Revivification"
+last_modified_at: 2022-02-11
+toc: true
+---
+
+## Revivification in FasterKV
+
+Revivification in FasterKV refers to reusing ("revivifying") Tombstoned (deleted) records. This minimizes log growth (and unused space) for high-delete scenarios.
+
+There are two forms of revivification:
+- **In-Chain**: Tombstoned records are left in the hash chain. A subsequent Upsert or RMW of the same key will revivify the record if it has sufficient value length.
+- **FreeList**: A free list is maintained, and records that are Tombstoned and are at the tail of the hash chain (pointed to directly from the hash table) are removed from the hash chain and kept in a free list (a binned (by power of 2), multiple-partition-per-bin set of circular buffers).
+
+## In-Chain
+If the FreeList is not active, Tombstoned records are left in the hash chain. A subsequent Upsert or RMW of the same key will revivify the record if it has sufficient value length.
+
+In-Chain revivification is always on:
+- A Tombstoned record may not be at the head of the hash chain.
+- It is possible to fail the CAS due to a concurrent insert
+
+## FreeList
+If `FasterKVSettings.MaxFreeRecordsInBin` is nonzero, this activates the freelist.
+
+When the FreeList is active and a record is deleted, if it is at the tail of the hash chain, is not locked, and its PreviousAddress points below BeginAddress, then it can be CASed (Compare-And-Swapped) out of the hash chain. If this succeeds, it is added to the freelist.
+
+### Design of FreeList
+The FreeList hierarchy consists of:
+- The FreeRecordPool, which maintains the bins, deciding which bin should be used for Enqueue and Dequeue.
+- Multiple FreeRecordBins, including one for oversize (where the length is too long to fit into the FreeRecordHeader)
```

Comment:
Can you mention the largest bin size here?

Reply:
added:
- The largest bin is of size 16 * 2^16. This is because each FreeRecord holds only a long for storage, and 48 bits of that must be used for the address, leaving only 16 for the size. However, we scale the bin sizes by 16 since that is the minimum size.